src/main/java/com/google/api/gax/grpc/RetryingCallable.java

```diff
@@ -29,226 +29,151 @@
+    RetryAlgorithm retryAlgorithm =
+        new RetryAlgorithm(
+            new GrpcExceptionRetryAlgorithm(), new ExponentialRetryAlgorithm(retrySettings, clock));
+    this.scheduler = new ScheduledRetryingExecutor<>(retryAlgorithm, scheduler);
   }
 
   @Override
   public ApiFuture<ResponseT> futureCall(RequestT request, CallContext context) {
-    RetryingResultFuture resultFuture = new RetryingResultFuture();
-    context = getCallContextWithDeadlineAfter(context, retryParams.getTotalTimeout());
-    Retryer retryer =
-        new Retryer(
-            request,
-            context,
-            resultFuture,
-            retryParams.getInitialRetryDelay(),
-            retryParams.getInitialRpcTimeout(),
-            null);
-    resultFuture.issueCall(request, context, retryer);
-    return resultFuture;
+    GrpcRetryCallable<RequestT, ResponseT> retryCallable =
+        new GrpcRetryCallable<>(callable, request, context);
+
+    RetryingFuture<ResponseT> retryingFuture = scheduler.createFuture(retryCallable);
+    retryCallable.setExternalFuture(retryingFuture);
+    retryCallable.call();
+
+    return retryingFuture;
   }
 
   @Override
   public String toString() {
     return String.format("retrying(%s)", callable);
   }
 
-  private class Retryer implements Runnable, ApiFutureCallback<ResponseT> {
-    private final RequestT request;
-    private final CallContext context;
-    private final RetryingResultFuture resultFuture;
-    private final Duration retryDelay;
-    private final Duration rpcTimeout;
-    private final Throwable savedThrowable;
-
-    private Retryer(
-        RequestT request,
-        CallContext context,
-        RetryingResultFuture resultFuture,
-        Duration retryDelay,
-        Duration rpcTimeout,
-        Throwable savedThrowable) {
-      this.request = request;
-      this.context = context;
-      this.resultFuture = resultFuture;
-      this.retryDelay = retryDelay;
-      this.rpcTimeout = rpcTimeout;
-      this.savedThrowable = savedThrowable;
-    }
-
-    @Override
-    public void run() {
-      if (context.getCallOptions().getDeadlineNanoTime() < clock.nanoTime()) {
-        if (savedThrowable == null) {
-          resultFuture.setException(
-              Status.DEADLINE_EXCEEDED
-                  .withDescription("Total deadline exceeded without completing any call")
-                  .asException());
-        } else {
-          resultFuture.setException(savedThrowable);
-        }
-        return;
-      }
-      CallContext deadlineContext = getCallContextWithDeadlineAfter(context, rpcTimeout);
-      resultFuture.issueCall(request, deadlineContext, this);
-    }
+  private static CallContext getCallContextWithDeadlineAfter(
+      CallContext oldCtx, Duration rpcTimeout) {
+    CallOptions oldOpt = oldCtx.getCallOptions();
+    CallOptions newOpt = oldOpt.withDeadlineAfter(rpcTimeout.getMillis(), TimeUnit.MILLISECONDS);
+    CallContext newCtx = oldCtx.withCallOptions(newOpt);
```

Comment:
Sorry I did not notice this in my previous review. To be consistent, we should avoid using such abbreviations (e.g. context vs ctx)